rust/kernel/net/mod.rs

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Network subsystem in rust for linux kernel.
+
+use core::mem;
+
+pub mod device;
+pub mod ethtool;
+pub mod netlink;
+pub mod rtnl;
+pub mod skbuff;
+
+#[doc(inline)]
+pub use macros::rtnl_link_ops;
+
+/// Determine if the Ethernet address is a multicast.
+///
+/// Return true if the address is a multicast address.
+/// By definition the broadcast address is also a multicast address.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// `addr` has to be a valid address pointer
+#[cfg(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)]
+pub unsafe fn is_multicast_ether_addr(addr: *const u8) -> bool {
+    // SAFETY: addr is a valid address pointer as above required, and has efficient unaligned access as KCONFIG stats
+    let a: u32 = unsafe { core::ptr::read_unaligned(addr as *const u32) };
+
+    if cfg!(target_endian = "big") {
+        (0x01 & (a >> (((mem::size_of::<u32>() as u32) * 8) - 8))) != 0
+    } else {
+        (0x01 & a) != 0
+    }
+}
+
+/// Determine if the Ethernet address is a multicast.
+///
+/// Return true if the address is a multicast address.
+/// By definition the broadcast address is also a multicast address.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// `addr` has to be a valid address pointer
+#[cfg(not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS))]
+pub unsafe fn is_multicast_ether_addr(addr: *const u8) -> bool {
+    let a: u16 = unsafe { *(addr as *const u16) };
+
+    if cfg!(target_endian = "big") {
+        (0x01 & (a >> (((mem::size_of::<u16>() as u16) * 8) - 8))) != 0
+    } else {
+        (0x01 & a) != 0
+    }
+}
+
+/// Determine if give Ethernet address is all zeros.
+/// Return true if the address is all zeroes.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// `addr` must point to a valid ethernet address and is 16-bit aligned
+#[cfg(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)]
+pub unsafe fn is_zero_ether_addr(addr: *const u8) -> bool {
+    // SAFETY: addr points to a valid ethernet address and is 16-bit aligned as required above.
+    unsafe { *(addr as *const u32) | (*((addr as usize + 4) as *const u16) as u32) == 0 }
```

Comment:
You can use .offset(n) or .add(n) and .cast::<T> here.

Reply:
[core::ptr::NonNull.cast](https://doc.rust-lang.org/stable/core/ptr/struct.NonNull.html#method.cast)? Never used it

Reply:
@bjorn3 can you look at it's current state?

Reply:
LGTM